File: internal/mock/context_navigation_fixture.go
package mock

import (
	"bytes"
	_ "embed"
	"fmt"
	"io"
	"time"

	"github.com/peasant-labs/peasant/internal/ingest"
	"github.com/peasant-labs/schema"
	"gopkg.in/yaml.v3"
)

//go:embed testdata/context_navigation.yaml
var contextNavigationFixtureYAML []byte

// The canonical mounted current-parent navigation fixture. See the YAML header
// for what the three stored sessions demonstrate. The Go mock provider and the
// real-binary visual smoke read the same file.
type contextNavigationTurnFixture struct {
	Index     int    `yaml:"index"`
	Role      string `yaml:"role"`
	EntryType string `yaml:"entryType"`
	Depth     int    `yaml:"depth"`
	Ref       string `yaml:"ref"`
	Content   string `yaml:"content"`
	Timestamp string `yaml:"timestamp"`
}

type contextNavigationEarlierFixture struct {
	State string                         `yaml:"state"`
	Turns []contextNavigationTurnFixture `yaml:"turns"`
}

type contextNavigationRelationshipFixture struct {
	Kind          string `yaml:"kind"`
	TargetState   string `yaml:"targetState"`
	TargetLocalID string `yaml:"targetLocalId"`
	Evidence      string `yaml:"evidence"`
}

type contextNavigationSessionFixture struct {
	ID                   string                                 `yaml:"id"`
	Project              string                                 `yaml:"project"`
	Harness              string                                 `yaml:"harness"`
	Model                string                                 `yaml:"model"`
	StartTime            string                                 `yaml:"startTime"`
	EndTime              string                                 `yaml:"endTime"`
	DurationMins         float64                                `yaml:"durationMins"`
	TotalTokens          int                                    `yaml:"totalTokens"`
	TokensIn             int                                    `yaml:"tokensIn"`
	TokensOut            int                                    `yaml:"tokensOut"`
	TurnCount            int                                    `yaml:"turnCount"`
	ToolCallCount        int                                    `yaml:"toolCallCount"`
	InputSubmissionCount *int64                                 `yaml:"inputSubmissionCount"`
	Turns                []contextNavigationTurnFixture         `yaml:"turns"`
	EarlierHistory       []contextNavigationEarlierFixture      `yaml:"earlierHistory"`
	Relationships        []contextNavigationRelationshipFixture `yaml:"relationships"`
}

type contextNavigationExpectedFixture struct {
	ContextLabel      string `yaml:"contextLabel"`
	StarterLabel      string `yaml:"starterLabel"`
	StarterLinkAction string `yaml:"starterLinkAction"`
	EarlierSummary    string `yaml:"earlierSummary"`
}

type contextNavigationFixture struct {
	Project  strikeProjectFixture             `yaml:"project"`
	Child    contextNavigationSessionFixture  `yaml:"child"`
	Source   contextNavigationSessionFixture  `yaml:"source"`
	Parent   contextNavigationSessionFixture  `yaml:"parent"`
	Expected contextNavigationExpectedFixture `yaml:"expected"`
}

var canonicalContextNavigationFixture = mustLoadContextNavigationFixture()

func mustLoadContextNavigationFixture() contextNavigationFixture {
	fixture, err := loadContextNavigationFixture(contextNavigationFixtureYAML)
	if err != nil {
		panic(fmt.Sprintf("embedded mounted current-parent navigation fixture could not be loaded because its shared payload is invalid at internal/mock/testdata/context_navigation.yaml during mock-provider initialization; the transcript navigation validation surface cannot start safely; fix the named fixture field and rebuild Peasant: %v", err))
	}
	return fixture
}

func loadContextNavigationFixture(source []byte) (contextNavigationFixture, error) {
	decoder := yaml.NewDecoder(bytes.NewReader(source))
	decoder.KnownFields(true)
	var fixture contextNavigationFixture
	if err := decoder.Decode(&fixture); err != nil {
		return contextNavigationFixture{}, fmt.Errorf("decode strict YAML: %w", err)
	}
	var extra any
	if err := decoder.Decode(&extra); err != io.EOF {
		return contextNavigationFixture{}, fmt.Errorf("expected exactly one YAML document, received trailing content")
	}

	projectHash, err := schema.NewProjectHash(fixture.Project.Hash)
	if err != nil {
		return contextNavigationFixture{}, fmt.Errorf("project.hash %q is invalid: %w", fixture.Project.Hash, err)
	}
	if projectHash != mockProjectHash(fixture.Project.Name) {
		return contextNavigationFixture{}, fmt.Errorf("project.hash %q does not match mock project %q", projectHash, fixture.Project.Name)
	}
	for name, session := range map[string]contextNavigationSessionFixture{
		"child":  fixture.Child,
		"source": fixture.Source,
		"parent": fixture.Parent,
	} {
		if _, err := schema.NewSessionID(session.ID); err != nil {
			return contextNavigationFixture{}, fmt.Errorf("%s.id %q is invalid: %w", name, session.ID, err)
		}
		if session.Project != fixture.Project.Name {
			return contextNavigationFixture{}, fmt.Errorf("%s.project %q must be %q", name, session.Project, fixture.Project.Name)
		}
		if session.TurnCount != len(session.Turns) {
			return contextNavigationFixture{}, fmt.Errorf("%s.turnCount %d must equal its %d turns", name, session.TurnCount, len(session.Turns))
		}
		if _, err := time.Parse(time.RFC3339, session.StartTime); err != nil {
			return contextNavigationFixture{}, fmt.Errorf("%s.startTime %q is invalid: %w", name, session.StartTime, err)
		}
		if _, err := time.Parse(time.RFC3339, session.EndTime); err != nil {
			return contextNavigationFixture{}, fmt.Errorf("%s.endTime %q is invalid: %w", name, session.EndTime, err)
		}
	}
	if fixture.Child.InputSubmissionCount == nil || *fixture.Child.InputSubmissionCount != 1 {
		return contextNavigationFixture{}, fmt.Errorf("child.inputSubmissionCount must be present and 1 for the one-submission fixture")
	}
	if fixture.Child.TurnCount != 5 {
		return contextNavigationFixture{}, fmt.Errorf("child.turnCount must be 5, received %d", fixture.Child.TurnCount)
	}
	if len(fixture.Child.EarlierHistory) != 1 || len(fixture.Child.EarlierHistory[0].Turns) == 0 {
		return contextNavigationFixture{}, fmt.Errorf("child.earlierHistory must carry exactly one non-empty retained partition")
	}
	if len(fixture.Child.Relationships) != 2 {
		return contextNavigationFixture{}, fmt.Errorf("child.relationships must carry the context_from source and the started_by parent, received %d", len(fixture.Child.Relationships))
	}
	for kind, want := range map[string]string{
		string(schema.SessionRelationshipContextFrom): fixture.Source.ID,
		string(schema.SessionRelationshipStartedBy):   fixture.Parent.ID,
	} {
		found := ""
		for _, relationship := range fixture.Child.Relationships {
			if relationship.Kind == kind {
				found = relationship.TargetLocalID
			}
		}
		if found != want {
			return contextNavigationFixture{}, fmt.Errorf("child relationship %q must target %q, received %q", kind, want, found)
		}
	}
	if fixture.Expected.ContextLabel == "" || fixture.Expected.StarterLabel == "" || fixture.Expected.StarterLinkAction == "" || fixture.Expected.EarlierSummary == "" {
		return contextNavigationFixture{}, fmt.Errorf("expected labels and summary must be non-empty")
	}
	return fixture, nil
}

func (fixture contextNavigationFixture) storedSessions() []ingest.Session {
	sessions := make([]ingest.Session, 0, 3)
	sessions = append(sessions, fixture.session(fixture.Child), fixture.session(fixture.Source), fixture.session(fixture.Parent))
	return sessions
}

func (fixture contextNavigationFixture) session(source contextNavigationSessionFixture) ingest.Session {
	startTime, _ := time.Parse(time.RFC3339, source.StartTime)
	endTime, _ := time.Parse(time.RFC3339, source.EndTime)
	session := ingest.Session{
		ID:        schema.SessionID(source.ID),
		Project:   source.Project,
		Harness:   schema.Harness(source.Harness),
		StartTime: startTime,
		EndTime:   endTime,
		Turns:     contextNavigationTurns(source.Turns),
		Model:     source.Model,
		Metadata: ingest.SessionMetadata{
			TokensIn:      source.TokensIn,
			TokensOut:     source.TokensOut,
			TotalTokens:   source.TotalTokens,
			Duration:      endTime.Sub(startTime),
			TurnCount:     source.TurnCount,
			ToolCallCount: source.ToolCallCount,
		},
		InputSubmissionCount: source.InputSubmissionCount,
	}
	for _, earlier := range source.EarlierHistory {
		session.EarlierHistory = append(session.EarlierHistory, ingest.EarlierHistorySection{
			State: schema.EarlierHistoryState(earlier.State),
			Turns: contextNavigationTurns(earlier.Turns),
		})
	}
	for _, relationship := range source.Relationships {
		target := schema.SessionID(relationship.TargetLocalID)
		session.Relationships = append(session.Relationships, schema.SessionRelationship{
			Kind:          schema.SessionRelationshipKind(relationship.Kind),
			TargetState:   schema.RelationshipTargetState(relationship.TargetState),
			TargetLocalID: &target,
			Evidence:      schema.EvidenceKind(relationship.Evidence),
		})
	}
	return session
}

func contextNavigationTurns(sources []contextNavigationTurnFixture) []ingest.Turn {
	turns := make([]ingest.Turn, len(sources))
	for index, source := range sources {
		timestamp, _ := time.Parse(time.RFC3339, source.Timestamp)
		turns[index] = ingest.Turn{
			Index:          source.Index,
			Role:           schema.Role(source.Role),
			Content:        source.Content,
			Timestamp:      timestamp,
			Depth:          source.Depth,
			EntryType:      schema.EntryType(source.EntryType),
			SourceEntryRef: source.Ref,
		}
	}
	return turns
}

File: web/scripts/visual/context-navigation-fixture.mjs
import { readFileSync } from 'node:fs'
import { dirname, resolve } from 'node:path'
import { fileURLToPath } from 'node:url'
import { newProjectHash } from '@peasant-labs/schema'
import { parseDocument } from 'yaml'

const HERE = dirname(fileURLToPath(import.meta.url))
const FIXTURE_PATH = resolve(HERE, '../../../internal/mock/testdata/context_navigation.yaml')
const SESSION_ID = /^sess?_[a-zA-Z0-9]+$/

function record(value, path) {
  if (!value || typeof value !== 'object' || Array.isArray(value)) throw new Error(`${path} must be a mapping`)
  return value
}

function exactFields(value, fields, path) {
  const keys = Object.keys(value).sort()
  const expected = [...fields].sort()
  if (keys.length !== expected.length || keys.some((key, index) => key !== expected[index])) {
    throw new Error(`${path} fields ${JSON.stringify(keys)} do not exactly match ${JSON.stringify(expected)}`)
  }
}

function requiredSessionId(value, field, path) {
  if (typeof value[field] !== 'string' || !SESSION_ID.test(value[field])) throw new Error(`${path}.${field} must be an owner-local session id`)
  return value[field]
}

function requiredString(value, field, path) {
  if (typeof value[field] !== 'string' || value[field].length === 0) throw new Error(`${path}.${field} must be a non-empty string`)
  return value[field]
}

/** The stored content of a fixture session's opening turn, read off the same file the Go store serves. */
function openingTurn(session, path) {
  const turns = session.turns
  if (!Array.isArray(turns) || turns.length === 0) throw new Error(`${path}.turns must carry at least the opening record`)
  return requiredString(record(turns[0], `${path}.turns[0]`), 'content', `${path}.turns[0]`)
}

/**
 * Loads the ONE mounted current-navigation fixture shared with the Go mock
 * provider, so the capture and the server can never disagree about the child
 * route, its exact context/source and parent targets, the unresolved reference,
 * or the labels the real viewer renders.
 */
export function loadContextNavigationFixture() {
  try {
    const document = parseDocument(readFileSync(FIXTURE_PATH, 'utf8'), { prettyErrors: true, strict: true, uniqueKeys: true })
    if (document.errors.length > 0) throw document.errors[0]
    if (document.contents === null) throw new Error('fixture document is empty')
    const root = record(document.toJS(), 'context navigation fixture')
    exactFields(root, ['project', 'child', 'source', 'parent', 'unresolved', 'expected'], 'context navigation fixture')

    const project = record(root.project, 'context navigation fixture.project')
    exactFields(project, ['hash', 'name'], 'context navigation fixture.project')
    const projectHash = newProjectHash(requiredString(project, 'hash', 'context navigation fixture.project'))
    const projectName = requiredString(project, 'name', 'context navigation fixture.project')

    const child = record(root.child, 'context navigation fixture.child')
    const source = record(root.source, 'context navigation fixture.source')
    const parent = record(root.parent, 'context navigation fixture.parent')
    const unresolved = record(root.unresolved, 'context navigation fixture.unresolved')
    const childSessionId = requiredSessionId(child, 'id', 'context navigation fixture.child')
    const sourceId = requiredSessionId(source, 'id', 'context navigation fixture.source')
    const parentId = requiredSessionId(parent, 'id', 'context navigation fixture.parent')
    const unresolvedChildId = requiredSessionId(unresolved, 'id', 'context navigation fixture.unresolved')
    const relationships = child.relationships
    if (!Array.isArray(relationships) || relationships.length !== 2) throw new Error('context navigation fixture.child.relationships must name the context_from source and the started_by parent')
    for (const relationship of relationships) {
      exactFields(record(relationship, 'context navigation fixture.child.relationships[]'), ['kind', 'targetState', 'targetLocalId', 'evidence'], 'context navigation fixture.child.relationships[]')
    }
    const context = relationships.find((relationship) => relationship.kind === 'context_from')
    const starter = relationships.find((relationship) => relationship.kind === 'started_by')
    if (context?.targetLocalId !== sourceId || starter?.targetLocalId !== parentId) throw new Error('context navigation fixture relationships must target the declared source and parent sessions')
    if (context?.targetState !== 'target_known' || starter?.targetState !== 'target_known' || context?.evidence !== 'native_typed' || starter?.evidence !== 'native_typed') throw new Error('context navigation fixture relationships must be known native-typed targets')

    // The unresolved child must name a target the store does NOT hold, so the
    // honest-reference arm can never accidentally become a link.
    const unresolvedRelationships = unresolved.relationships
    if (!Array.isArray(unresolvedRelationships) || unresolvedRelationships.length !== 1) throw new Error('context navigation fixture.unresolved.relationships must carry the one started_by relationship')
    const absentTargetId = requiredSessionId(record(unresolvedRelationships[0], 'context navigation fixture.unresolved.relationships[0]'), 'targetLocalId', 'context navigation fixture.unresolved.relationships[0]')
    if ([childSessionId, sourceId, parentId, unresolvedChildId].includes(absentTargetId)) throw new Error(`context navigation fixture.unresolved relationship target ${absentTargetId} names a stored session; the unavailable reference case requires an absent target`)

    const expected = record(root.expected, 'context navigation fixture.expected')
    exactFields(expected, ['contextLabel', 'starterLabel', 'starterLinkAction', 'earlierSummary'], 'context navigation fixture.expected')

    return Object.freeze({
      projectHash,
      projectName,
      childSessionId,
      childOpening: openingTurn(child, 'context navigation fixture.child'),
      sourceId,
      sourceOpening: openingTurn(source, 'context navigation fixture.source'),
      parentId,
      parentOpening: openingTurn(parent, 'context navigation fixture.parent'),
      unresolvedChildId,
      absentTargetId,
      contextLabel: requiredString(expected, 'contextLabel', 'context navigation fixture.expected'),
      starterLabel: requiredString(expected, 'starterLabel', 'context navigation fixture.expected'),
      starterLinkAction: requiredString(expected, 'starterLinkAction', 'context navigation fixture.expected'),
      earlierSummary: requiredString(expected, 'earlierSummary', 'context navigation fixture.expected'),
    })
  } catch (error) {
    throw new Error(
      `Mounted context navigation fixture could not be loaded because ${error.message} at ${FIXTURE_PATH} before the capture server started; the mounted link/Back evidence would not match the served data; fix the fixture structure or schema values, then rerun context-navigation-shoot.mjs.`,
      { cause: error },
    )
  }
}
